---
aliases: Error.prototype.cause
---

#Errorprototypecause


#### Error.prototype.cause
一个Error实例的`cause`数据属性表明错误最初的具体原因.(specific original cause of error)
当捕获或重新抛出带有更具体或有用的错误信息的错误,以便仍能使用最初的错误时使用.



**属性特性**

| Name         | value                                           |
| ------------ | ----------------------------------------------- |
| cause        | 在`options.cause`参数中传递给Error构造函数的值.可能没声明 |
| Writable     | yes                                             |
| Enumerable   | No                                              |
| Configurable | yes                                             |




**描述**
`cause`的值可以是任意类型. 不应假设捕获的错误拥有一个`Error`作为它的`cause`, 同样情况下也不能保证`catch`语句绑定的变量是一个`Error`.

以下提供结构化数据作为错误原因(error cause)案例,展示non-error被谨慎提供作为原因的案例.

**案例**
Rethrowing an error with a cause
```js
try {
	connectToDatabase()
} catch(err) {
	throw new Error('Connecting to database failed.', {cause: err})
}
```
上面catch捕获的err是Error的子类[[202301302098a3|ReferenceError]]


提供结构化数据作为error cause
为人类阅读编写的错误信息,可能对机器解析不太合适. 因为很容易受到重写或标点变动的影响,这可能会破坏为理解它们而编写的现有解析.  ???
所以从函数中抛出错误时, 作为一个人类可读错误信息的代替,你可以为机器解析,提供结构化数据的原因.

```js
function makeRSA(p, q) {
  if (!Number.isInteger(p) || !Number.isInteger(q)) {
    throw new Error('RSA key generation requires integer inputs.', {
      cause: { code: 'NonInteger', values: [p, q] },
    });
  }
  if (!areCoprime(p, q)) {
    throw new Error('RSA key generation requires two co-prime integers.', {
      cause: { code: 'NonCoprime', values: [p, q] },
    })
  }
  // rsa algorithm…
}
```




#### 来源
[MDN Error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause)